msg-socket/src/pub/session.rs

```diff
@@ -0,0 +1,196 @@
+use futures::{Future, SinkExt, StreamExt};
+use std::{
+    borrow::Cow,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+use tokio::io::{AsyncRead, AsyncWrite};
+use tokio_stream::wrappers::BroadcastStream;
+use tokio_util::codec::Framed;
+use tracing::{debug, error, trace, warn};
+
+use super::{trie::PrefixTrie, PubMessage, SocketState};
+use msg_wire::pubsub;
+
+pub(super) struct SubscriberSession<Io> {
+    /// The sequence number of this session.
+    pub(super) seq: u32,
+    /// The ID of this session.
+    pub(super) session_id: u32,
+    /// Messages from the socket.
+    pub(super) from_socket_bcast: BroadcastStream<PubMessage>,
+    /// Messages queued to be sent on the connection
+    pub(super) pending_egress: Option<pubsub::Message>,
+    /// Session request sender.
+    // to_driver: mpsc::Sender<SessionRequest>,
+    pub(super) state: Arc<SocketState>,
+    /// The framed connection.
+    pub(super) conn: Framed<Io, pubsub::Codec>,
+    /// The topic filter (a prefix trie that works with strings)
+    pub(super) topic_filter: PrefixTrie,
+    pub(super) flush_interval: Option<tokio::time::Interval>,
+}
+
+impl<Io: AsyncRead + AsyncWrite + Unpin> SubscriberSession<Io> {
+    #[inline]
+    fn on_outgoing(&mut self, msg: PubMessage) {
+        // Check if the message matches the topic filter
+        if self.topic_filter.contains(msg.topic()) {
+            trace!(
+                topic = msg.topic(),
+                "Message matches topic filter, adding to egress queue"
+            );
+
+            // Generate the wire message and increment the sequence number
+            self.pending_egress = Some(msg.into_wire(self.seq));
+            self.seq = self.seq.wrapping_add(1);
+        }
+    }
+
+    #[inline]
+    fn on_incoming(&mut self, msg: pubsub::Message) {
+        // The only incoming messages we should have are control messages.
+        match msg_to_control(&msg) {
+            ControlMsg::Subscribe(topic) => {
+                debug!("Subscribing to topic {:?}", topic);
+                self.topic_filter.insert(&topic)
+            }
+            ControlMsg::Unsubscribe(topic) => {
+                debug!("Unsubscribing from topic {:?}", topic);
+                self.topic_filter.remove(&topic)
+            }
+            ControlMsg::Close => todo!(),
```

Comment:
how are session closes handled? If the `conn` yields None then the future stops and that's it?

Reply:
https://github.com/chainbound/msg-rs/blob/7b3ab0b29c82bfe42496f4bbccc4feefd82ce3b2/msg-socket/src/pub/session.rs#L183-L189

Correct. This results in a `Drop` of the `SubscriberSession`, which removes all memory associated to that session

Reply:
Awesome, thanks, do you think it's worth removing the todo!() in line 63 then?